apps/resource_app.go

```diff
@@ -0,0 +1,181 @@
+	// provides a system generated stable snapshotted source code path used by
+	// the deployment.
+	SourceCodePath string `json:"source_code_path,omitempty"`
+}
+
+type appCreateStruct struct {
+	apps.CreateAppRequest
+}
+
+func (appCreateStruct) Aliases() map[string]map[string]string {
+	return map[string]map[string]string{
+		"apps.appCreateStruct": appAliasMap,
+	}
+}
+
+func (appCreateStruct) CustomizeSchema(s *common.CustomizableSchema) *common.CustomizableSchema {
+	return s
+}
+
+type appUpdateStruct struct {
+	apps.UpdateAppRequest
+}
+
+func (appUpdateStruct) Aliases() map[string]map[string]string {
+	return map[string]map[string]string{
+		"apps.appUpdateStruct": appAliasMap,
+	}
+}
+
+func (appUpdateStruct) CustomizeSchema(s *common.CustomizableSchema) *common.CustomizableSchema {
+	return s
+}
+
+func (appStruct) Aliases() map[string]map[string]string {
+	return map[string]map[string]string{
+		"apps.appStruct": appAliasMap,
+	}
+}
+
+func (appStruct) CustomizeSchema(s *common.CustomizableSchema) *common.CustomizableSchema {
+
+	// Required fields & validation
+	s.SchemaPath("name").SetRequired().SetForceNew().SetValidateFunc(validation.StringLenBetween(2, 30))
+	s.SchemaPath("mode").SetDefault("SNAPSHOT").SetValidateFunc(validation.StringInSlice([]string{"SNAPSHOT", "AUTO_SYNC"}, false))
+
+	// Computed fields
+	for _, p := range []string{"active_deployment", "app_status", "compute_status", "create_time", "creator",
+		"default_source_code_path", "pending_deployment", "service_principal_id", "service_principal_name",
+		"update_time", "updater", "url"} {
+		s.SchemaPath(p).SetComputed()
+	}
+
+	// SuppressDiff
+	s.SchemaPath("source_code_path").SetCustomSuppressDiff(common.WorkspacePathPrefixDiffSuppress)
+
+	return s
+}
+
+var appSchema = common.StructToSchema(appStruct{}, nil)
+
+func ResourceApp() common.Resource {
+	return common.Resource{
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			var createApp appCreateStruct
+			var createAppDeployment apps.CreateAppDeploymentRequest
+			common.DataToStructPointer(d, appSchema, &createApp)
+			common.DataToStructPointer(d, appSchema, &createAppDeployment)
+			w, err := c.WorkspaceClient()
+			if err != nil {
+				return err
+			}
+
+			// create the app, which does not require the source code path yet
+			wait, err := w.Apps.Create(ctx, createApp.CreateAppRequest)
+			if err != nil {
+				return err
+			}
+			app, err := wait.GetWithTimeout(d.Timeout(schema.TimeoutCreate) - deleteCallTimeout)
+			if err != nil {
+				log.Printf("[ERROR] Error waiting for app to be created: %s", err.Error())
+				_, nestedErr := w.Apps.DeleteByName(ctx, createApp.Name)
+				if nestedErr != nil {
+					log.Printf("[ERROR] Error cleaning up app: %s", nestedErr.Error())
+				}
+				return err
+			}
+			// now deploy the app, using the source code path
+			createAppDeployment.AppName = app.Name
```

Comment:
I think we discussed it offline and agreed that the deployment won't be part of TF configuration. Instead, if an users want to deploy, they use CLI to do that. Otherwise it's confusing because we do the deployments only when source_code_path changed but we need to do deployments when source code is changed.

@pietern wdyt?